lsp/nls/src/requests/completion.rs

```diff
@@ -285,6 +333,68 @@ mod tests {
     use super::*;
     use std::collections::{HashMap, HashSet};
 
+    #[test]
+    fn test_get_identifier_path() {
+        let tests = [
+            ("let person = {} in \n    a.b.c.d", vec!["a", "b", "c", "d"]),
+            ("   ab.dec", vec!["ab", "dec"]),
+            (
+                r##"let name = { sdf.clue.add.bar = 10 } in
+            name.sdf.clue.add.bar"##,
+                vec!["name", "sdf", "clue", "add", "bar"],
+            ),
+            ("name.class", vec!["name", "class"]),
+            ("name.class.", vec!["name", "class"]),
+            ("number", vec!["number"]),
+            (
+                r##"let x = {"fo京o" = {bar = 42}} in x."fo京o".foo"##,
+                vec!["x", "\"fo京o\"", "foo"],
+            ),
+            (
+                r##"let me : _ = { name = "gg", age = 23, a.b.c.d = 10 } in
+                    me.ca.cb"##,
+                vec!["me", "a", "b"],
+            ),
+        ];
+        for (input, expected) in tests {
+            let actual = get_identifier_path(input);
+            let expected: Option<Vec<_>> =
+                Some(expected.iter().cloned().map(String::from).collect());
+            assert_eq!(actual, expected)
+        }
+    }
+
+    #[test]
+    fn test_extract_ident_with_path() {
+        let ty = Types(AbsType::RowExtend(
```

Comment:
We could potentially use a macro (similar to `mk_tyw_row`) here to simplify this construction. Totally up to you though, as the duplication only appears in one place currently.

Reply:
Yes, I plan to use macros to abstract a lot of the tests in the future.